### PR TITLE
add shutdown hook to singularity executor

### PR DESCRIPTION
**Pull request by ssalinas**
right now SingularityExecutor isn't gracefully shutting down when it receives a signal. Adding a shutdown hook to trigger the same shutdown process (`monitor.shutdown()`) as would happen when mesos tells it to shut down

### CODE REVIEW

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/SingularityExecutorRunner.java [hunk 1]
```diff
@@ -54,6 +58,16 @@ public SingularityExecutorRunner(@Named(SingularityRunnerBaseModule.PROCESS_NAME
 
     final MesosExecutorDriver driver = new MesosExecutorDriver(singularityExecutor);
 
+    Runtime.getRuntime().addShutdownHook(new Thread("SingularityExecutorRunnerGracefulShutdown") {
+
+      @Override
+      public void run() {
+        LOG.info("Executor is shutting down...");
```
Comment (wsorenson) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/SingularityExecutorRunner.java]: Why is it shutting down? Always be specific :)

**Comment by ssalinas** [comment]
@tpetr @wsorenson the lock on shutdown works the way we thought it would. No processes left hanging waiting for locks

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/SingularityExecutorMonitor.java [hunk 2]
```diff
@@ -87,7 +91,16 @@ public SingularityExecutorMonitor(SingularityExecutorLogging logging, ExecutorUt
     SUBMITTED, REJECTED, TASK_ALREADY_EXISTED;
   }
 
-  public void shutdown(Optional<ExecutorDriver> driver) {
+  public void shutdownWithLock(Optional<ExecutorDriver> driver) {
+    try {
+      shutdownLock.lock();
```
Comment (tpetr) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/SingularityExecutorMonitor.java]: consider using a timeout

**Comment by ssalinas** [comment]
@tpetr tried one last thing, this version of things will shut down cleanly every time that I have tried now, need to test more still though

**Pull request merged**